### PR TITLE
Refactor of #8 to use paramethers and headers

### PR DESCRIPTION
**Pull request by PingIsFun**
Forked this from https://github.com/HoggyTheWizard/hypixel-api-python/tree/27f8512abe825ad30abaf76abbfd3c1443158ba6 because #8 wasn't merged yet.

Refactor of #8 to use paramethers and headers
Added a `force` option to purge cache.
Added return types to some functions.

TODO:

- [ ] annotate all function arguments and return types

### CODE REVIEW

File: src/hypxiel_api_python/utils/requests.py [hunk 1]
```diff
@@ -1,34 +1,63 @@
...
-async def get(self, endpoint: str, param: tuple, nocache: bool = None):
-    api = self.hypixel_api
+async def get(hypixel_api: HypixelAPI, endpoint: str, params: dict[str, str], nocache: bool = None) -> dict:
     """Determine whether it should use a cached object or fetch a new one."""
     """`nocache` is a way for the user to bypass the cache"""
-    cached_value = self.hypixel_api.endpoints[param][endpoint.replace("/", "")]
-    if not cached_value:
-        return await fetch(self=self, url=f"{api.path}/{endpoint}?key={api.key}&{param[0]}={param[1]}")
+    if not endpoint.startswith("/"):
+        endpoint = f"/{endpoint}"
```
Comment (ChiefChippy2) [inline src/hypxiel_api_python/utils/requests.py]: is this really necessary?

File: src/hypxiel_api_python/utils/requests.py [hunk 2]
```diff
@@ -1,34 +1,63 @@
...
-        return await fetch(self=self, url=f"{api.path}/{endpoint}?key={api.key}&{param[0]}={param[1]}")
+    await purge_cache(hypixel_api)
+    auth_header = {"API-Key": hypixel_api.key}
+    try:
+        cached_response = hypixel_api.request_cache[f"{endpoint}.{params}"]
+        # print(cached_response)
+    except KeyError:
+        cached_response = None
+    # if not cached_response:
+    #     return await fetch(self, endpoint, params, auth_header)
```
Comment (ChiefChippy2) [inline src/hypxiel_api_python/utils/requests.py]: alternatively use cached_response= ...request_cache.get("key")

File: src/hypxiel_api_python/utils/requests.py [hunk 3]
```diff
@@ -1,34 +1,63 @@
...
+    if not cached_response or cached_response["hypixel-api-python"]["timestamp"] + 60 > int(datetime.now().timestamp()) or nocache:
```
Comment (ChiefChippy2) [inline src/hypxiel_api_python/utils/requests.py]: please don't hardcode cache expiration time...

File: src/hypxiel_api_python/utils/requests.py [hunk 4]
```diff
@@ -1,34 +1,63 @@
...
+    :return: A dictionary containing the response from the Hypixel API.
+    """
+    HYPIXEL_API_URL = "https://api.hypixel.net"
 
-            # ratelimited
+    async with ClientSession(HYPIXEL_API_URL, headers=headers) as session:
+        async with session.get(endpoint, params=params) as request:
+            ratelimit_info = {x: y for x, y in request.headers.items() if x.startswith("RateLimit-")}
+            # ratelimit
+            print(request.status)
```
Comment (ChiefChippy2) [inline src/hypxiel_api_python/utils/requests.py]: this isn't production yet but remember to remove this xd

**Review by ChiefChippy2 — COMMENTED** [review]
Check my comments

File: src/hypxiel_api_python/utils/requests.py (hunk 1 above)
Comment (PingIsFun) [reply src/hypxiel_api_python/utils/requests.py]: not really

Comment (PingIsFun) [reply src/hypxiel_api_python/utils/requests.py]: Was also thinking about using an enum for endpoint but idk. is that a good idea?

Comment (ChiefChippy2) [reply src/hypxiel_api_python/utils/requests.py]: sure ig... but it's out of the scope of my comment :)

**Comment by HoggyTheWizard** [comment]
I don't like `fetch` being renamed to `fetch_hypixel_api` and we'll have to discuss how we're going to handle documentation as the docstrings are a little intrusive. Other than that it looks good.

**Comment by Netwyn** [comment]
Looks good

**Pull request merged**